src/Lean/Elab/Tactic/NormCast.lean

```diff
@@ -0,0 +1,254 @@
+/-
+Copyright (c) 2019 Paul-Nicolas Madelaine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul-Nicolas Madelaine, Robert Y. Lewis, Mario Carneiro, Gabriel Ebner
+-/
+import Lean.Meta.Tactic.NormCast
+import Lean.Elab.Tactic.Conv.Simp
+import Lean.Elab.ElabRules
+
+/-!
+# The `norm_cast` family of tactics.
+-/
+namespace Lean.Elab.Tactic.NormCast
+open Lean Meta Simp NormCast
+
+-- TODO: trace name consistency
+builtin_initialize registerTraceClass `Tactic.norm_cast
+
+/-- Prove `a = b` using the given simp set. -/
+def proveEqUsing (s : SimpTheorems) (a b : Expr) : MetaM (Option Simp.Result) := do
+  let go : SimpM (Option Simp.Result) := do
+    let a' ← Simp.simp a
+    let b' ← Simp.simp b
+    unless ← isDefEq a'.expr b'.expr do return none
+    a'.mkEqTrans (← b'.mkEqSymm b)
+  withReducible do
+    (go (← Simp.mkDefaultMethods).toMethodsRef
+      { simpTheorems := #[s], congrTheorems := ← Meta.getSimpCongrTheorems }).run' {}
+
+/-- Prove `a = b` by simplifying using move and squash lemmas. -/
+def proveEqUsingDown (a b : Expr) : MetaM (Option Simp.Result) := do
+  withTraceNode `Tactic.norm_cast (return m!"{exceptOptionEmoji ·} proving: {← mkEq a b}") do
+  proveEqUsing (← normCastExt.down.getTheorems) a b
+
+/-- Construct the expression `(e : ty)`. -/
+def mkCoe (e : Expr) (ty : Expr) : MetaM Expr := do
+  let .some e' ← coerce? e ty | failure
+  return e'
+
+/--
+Check if an expression is the coercion of some other expression,
+and if so return that expression.
+-/
+def isCoeOf? (e : Expr) : MetaM (Option Expr) := do
+  if let Expr.const fn .. := e.getAppFn then
+    if let some info ← getCoeFnInfo? fn then
+      if e.getAppNumArgs == info.numArgs then
+        return e.getArg! info.coercee
+  return none
+
+/--
+Check if an expression is a numeral in some type,
+and if so return that type and the natural number.
+-/
+def isNumeral? (e : Expr) : Option (Expr × Nat) :=
+  -- TODO: cleanup, and possibly remove duplicate
+  if e.isConstOf ``Nat.zero then
+    (mkConst ``Nat, 0)
+  else if let Expr.app (Expr.app (Expr.app (Expr.const ``OfNat.ofNat ..) α ..)
+      (Expr.lit (Literal.natVal n) ..) ..) .. := e then
+    some (α, n)
+  else
+    none
+
+/--
+This is the main heuristic used alongside the elim and move lemmas.
+The goal is to help casts move past operators by adding intermediate casts.
+An expression of the shape: op (↑(x : α) : γ) (↑(y : β) : γ)
+is rewritten to:            op (↑(↑(x : α) : β) : γ) (↑(y : β) : γ)
+when (↑(↑(x : α) : β) : γ) = (↑(x : α) : γ) can be proven with a squash lemma
```

Comment:
````suggestion
An expression of the shape:
```
op (↑(x : α) : γ) (↑(y : β) : γ)
```
is rewritten to:
```
op (↑(↑(x : α) : β) : γ) (↑(y : β) : γ)
```
when
```
(↑(↑(x : α) : β) : γ) = (↑(x : α) : γ)
```
can be proven with a squash lemma
````
Should be easier to read when rendered.